src/Microsoft.CodeAnalysis.Analyzers/Core/MetaAnalyzers/SourceGeneratorAttributeAnalyzer.cs

```diff
@@ -0,0 +1,69 @@
+        private static readonly LocalizableString s_localizableMessage = new LocalizableResourceString(nameof(CodeAnalysisDiagnosticsResources.MissingSourceGeneratorAttributeTitle), CodeAnalysisDiagnosticsResources.ResourceManager, typeof(CodeAnalysisDiagnosticsResources));
+        private static readonly LocalizableString s_localizableDescription = new LocalizableResourceString(nameof(CodeAnalysisDiagnosticsResources.MissingSourceGeneratorAttributeDescription), CodeAnalysisDiagnosticsResources.ResourceManager, typeof(CodeAnalysisDiagnosticsResources));
+
+        public static readonly DiagnosticDescriptor DiagnosticRule = new(
+            DiagnosticIds.MissingSourceGeneratorAttributeId,
+            s_localizableTitle,
+            s_localizableMessage,
+            DiagnosticCategory.MicrosoftCodeAnalysisCorrectness,
+            DiagnosticSeverity.Warning,
+            description: s_localizableDescription,
+            isEnabledByDefault: true);
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(DiagnosticRule);
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.EnableConcurrentExecution();
+
+            context.RegisterCompilationStartAction(compilationContext =>
+            {
+                var sourceGenerator = compilationContext.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftCodeAnalysisISourceGenerator);
+                var sourceGeneratorAttribute = compilationContext.Compilation.GetOrCreateTypeByMetadataName(WellKnownTypeNames.MicrosoftCodeAnalysisGeneratorAttribute);
+
+                if (sourceGenerator == null || sourceGeneratorAttribute == null)
+                    // We don't need to check assemblies unless they're referencing Microsoft.CodeAnalysis which defines DiagnosticAnalyzer.
+                    return;
+
+                compilationContext.RegisterSymbolAction(c => AnalyzeSymbol(c, sourceGenerator, sourceGeneratorAttribute), SymbolKind.NamedType);
+            });
+        }
+
+        private static void AnalyzeSymbol(SymbolAnalysisContext c, INamedTypeSymbol sourceGenerator, INamedTypeSymbol sourceGeneratorAttribute)
+        {
+            var symbol = (INamedTypeSymbol)c.Symbol;
+
+            if (symbol.IsAbstract || symbol.IsAnonymousType)
+            {
+                return;
+            }
+
+            if (!symbol.AllInterfaces.Any(i => sourceGenerator.Equals(i, SymbolEqualityComparer.Default)))
+            {
+                return;
+            }
+
+            if (symbol.GetApplicableAttributes(null).Any(a => a.AttributeClass.Inherits(sourceGeneratorAttribute)))
+            {
+                return;
+            }
```

Comment:
Is there any test hitting this condition?

Reply:
The generator attribute is an inherited attribute, so still technically correct afaik if it's in an ancestor class. If we really wanted to enforce otherwise we could make that change at the attribute definition. 

Correct that `TestHierarchy_InheritedAttribute` tests this code path.